### PR TITLE
[Intel HPU] enable level 1 prefix caching and fix some bugs

### PR DESCRIPTION
**Pull request by fmiao2372**
## Motivation
Develop new features for Intel HPU and fix its some bugs

## Modifications
Enable level 1 prefix caching(only HPU memory) on HPU
Use new fused_block_attention OP
Fix some bugs

The relevant patches have been merged into PaddleCustomizable Device

## Usage or Command
Follow this spec
https://github.com/PaddlePaddle/FastDeploy/blob/develop/docs/get_started/installation/intel_gaudi.md
add --enable-prefix-caching/--no-enable-prefix-caching to turn on/off prefix caching

## Accuracy Tests
ERNIE-4.5-21B-A3B-Paddle
100%|██████████| 1319/1319 [08:49<00:00,  2.49it/s]
Accuracy: 0.915
Invalid: 0.000
Latency: 529.256 s

ERNIE-4.5-300B-A47B-Paddle
100%|██████████| 1319/1319 [09:21<00:00,  2.35it/s]
Accuracy: 0.956
Invalid: 0.000
Latency: 561.466 s

## Checklist

- [ Done] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [Done ] Format your code, run `pre-commit` before commit.
- [ Done] Add unit tests. Please write the reason in this PR if no unit tests.
- Due to the integration of PaddleCustom device and the difficulty in separating individual units, we have conducted sufficient performance and accuracy testing locally
- [Done ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/worker/worker_process.py [hunk 1]
```diff
@@ -879,6 +884,21 @@ def initialize_fd_config(args, ranks: int = 1, local_rank: int = 0) -> FDConfig:
...
+    if parallel_config.enable_dense_tp_and_moe_ep:
+        expert_parallel_rank = int(local_rank % parallel_config.tensor_parallel_size)
+        if isinstance(model_config.moe_num_experts, list):
+            num_experts = model_config.moe_num_experts[0]
+        else:
+            num_experts = model_config.moe_num_experts
+
+        num_experts_per_rank = num_experts // parallel_config.tensor_parallel_size
+        num_experts_start_offset = expert_parallel_rank * num_experts_per_rank
+        max_chips_per_node = 16 if current_platform.is_iluvatar() else 8
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/worker/worker_process.py]: fixed

File: fastdeploy/model_executor/layers/backends/intel_hpu/attention/hpu_attn_backend.py [hunk 2]
```diff
@@ -297,11 +347,14 @@ def forward_decode(
             qkv_proj.weight,
             qkv_proj.bias,
             o_proj.weight,
+            None,
+            None,
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/model_executor/layers/backends/intel_hpu/attention/hpu_attn_backend.py]: fixed

File: fastdeploy/model_executor/load_weight_utils.py [hunk 3]
```diff
@@ -45,6 +45,62 @@
 from fastdeploy.platforms import current_platform
 
 
+def reload_ep_checkpoint(model_path: str, fd_config: FDConfig, state_dict: dict, return_numpy: bool = False):
+    """
+    load ep checkpoint
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/model_executor/load_weight_utils.py]: fixed

File: fastdeploy/worker/hpu_model_runner.py [hunk 4]
```diff
@@ -1026,7 +1037,7 @@ def _dummy_sampler_run(self) -> paddle.Tensor:
         """ """
         pass
 
-    def update_warmup_inputs(self, requests, is_decode=False):
+    def update_warmup_inputs(self, requests, is_decode=False, context_len=0) -> None:
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/worker/hpu_model_runner.py:1040]: fixed

File: fastdeploy/model_executor/layers/moe/moe.py [hunk 5]
```diff
@@ -149,9 +149,14 @@ def __init__(
         self.hidden_size = fd_config.model_config.hidden_size
         self.num_experts = num_experts
 
-        self.num_local_experts = self.num_experts // self.ep_size
+        if fd_config.parallel_config.enable_dense_tp_and_moe_ep:
+            self.num_local_experts = self.num_experts // self.tp_size
+            self.moe_intermediate_size = moe_intermediate_size
+            expert_id_offset = fd_config.parallel_config.tensor_parallel_rank * self.num_local_experts
```
[the remark this answers is not in the document]
Comment (fmiao2372) [reply fastdeploy/model_executor/layers/moe/moe.py]: misreport

File: fastdeploy/model_executor/layers/moe/moe.py [hunk 6]
```diff
@@ -149,9 +149,14 @@ def __init__(
         self.hidden_size = fd_config.model_config.hidden_size
         self.num_experts = num_experts
 
-        self.num_local_experts = self.num_experts // self.ep_size
+        if fd_config.parallel_config.enable_dense_tp_and_moe_ep:
+            self.num_local_experts = self.num_experts // self.tp_size
```
Comment (gzy19990617) [inline fastdeploy/model_executor/layers/moe/moe.py]: 这个是指按tp分配专家？

Comment (fmiao2372) [reply fastdeploy/model_executor/layers/moe/moe.py]: 对的，dense_tp_and_moe_ep这种会按照tp size来切分

File: fastdeploy/engine/args_utils.py [hunk 7]
```diff
@@ -829,6 +834,12 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             default=EngineArgs.enable_expert_parallel,
             help="Enable expert parallelism.",
         )
+        parallel_group.add_argument(
+            "--enable-dense-tp-and-moe-ep",
```
Comment (yuanlehome) [inline fastdeploy/engine/args_utils.py]: 我理解不需要添加这个参数，FD里的TP/EP混合并行中，默认MoE模块TP=1的，这个代码在 https://github.com/PaddlePaddle/FastDeploy/blob/develop/fastdeploy/model_executor/layers/moe/moe.py#L142 ，正常启动fastdeploy时配置tp>1 & ep > 1，会有啥问题吗？

File: fastdeploy/model_executor/load_weight_utils.py [hunk 8]
```diff
@@ -41,10 +41,79 @@
 from fastdeploy.model_executor.models.tp_utils import (
     check_tensor_parallel_prerequisites,
 )
-from fastdeploy.model_executor.utils import multi_switch_config_context
+from fastdeploy.model_executor.utils import switch_config_context
 from fastdeploy.platforms import current_platform
 
 
+def reload_ep_checkpoint(model_path: str, fd_config: FDConfig, state_dict: dict, return_numpy: bool = False):
```
Comment (yuanlehome) [inline fastdeploy/model_executor/load_weight_utils.py]: 请尽量在V1 loader下去支持，V0 loader不建议再开发了

**Comment by yuanlehome** [comment]
请完善PR描述～

File: fastdeploy/model_executor/layers/moe/moe.py (hunk 6 above)
Comment (yuanlehome) [reply fastdeploy/model_executor/layers/moe/moe.py]: 为什么不是按照ep_size来划分专家呢？

File: fastdeploy/engine/args_utils.py (hunk 7 above)
Comment (fmiao2372) [reply fastdeploy/engine/args_utils.py]: https://github.com/PaddlePaddle/FastDeploy/blob/develop/fastdeploy/config.py#L566
https://github.com/PaddlePaddle/FastDeploy/blob/develop/fastdeploy/model_executor/layers/moe/moe.py#L146
目前启动参数里面好像不能直接设置TP, 而且MoE里面TP和EP不能同时设置，EP是按照data_parallel_size来切分的

File: fastdeploy/model_executor/layers/moe/moe.py (hunk 6 above)
Comment (fmiao2372) [reply fastdeploy/model_executor/layers/moe/moe.py]: 同下，EP是按照data_parallel_size来切分的

File: fastdeploy/model_executor/load_weight_utils.py (hunk 8 above)
Comment (fmiao2372) [reply fastdeploy/model_executor/load_weight_utils.py]: 代码改动支持V1 loader的

**Comment by fmiao2372** [comment]
> 请完善PR描述～

已更新

File: fastdeploy/engine/args_utils.py (hunk 7 above)
Comment (yuanlehome) [reply fastdeploy/engine/args_utils.py]: 1. 启动参数里面这些都是可以同时设置的：

    - --tensor-parallel-size x
    - --enable-expert-parallel
    - --data-parallel-size y

2. 你们的EP跟GPU的EP或者说一般意义上大家理解的EP是有啥区别吗？为啥是根据data_parallel_size来切分？

File: fastdeploy/model_executor/load_weight_utils.py (hunk 8 above)
Comment (yuanlehome) [reply fastdeploy/model_executor/load_weight_utils.py]: V1 loader不会走到此函数，走的是xxx_model.py里的load_weights函数

File: fastdeploy/engine/args_utils.py (hunk 7 above)
Comment (fmiao2372) [reply fastdeploy/engine/args_utils.py]: 比如说一机八卡，需求是Attention部分走TP，MoE部分走EP，因为FD里面目前没有设置EP的参数，而且走EP必须要用DP，而我们不走DP，因此加一个enable这样的参数, MoE按照TP size来切就行

Comment (yuanlehome) [reply fastdeploy/engine/args_utils.py]: 1. 再仔细看一下上句的回答，FD支持设置EP
2. 走EP也并不要求DP > 1

File: fastdeploy/model_executor/load_weight_utils.py (hunk 8 above)
Comment (fmiao2372) [reply fastdeploy/model_executor/load_weight_utils.py]: 测试的时候load_choice改成了default_v1是好的，以为支持了。重新看了一下代码，针对非GPU/XPU都会fallback回default，我们目前确实不支持，后续会统一改一下的

Comment (yuanlehome) [reply fastdeploy/model_executor/load_weight_utils.py]: 期望这次能够一并修改了，在V1 loader下去支持，当前的实现强制与ernie 模型绑定是不合理的

File: fastdeploy/engine/args_utils.py (hunk 7 above)
Comment (LeoZhao-BJ) [reply fastdeploy/engine/args_utils.py]: 如果要支持attention部分走TP， MoE部分走EP 这种方式的执行, 启动参数如何设置 ？现在好像没有单独给attention/moe各自设置TP/EP size的参数吧？

        assert (self.tp_size >= 1 and self.ep_size == 1) or (
            self.tp_size == 1 and self.ep_size > 1
        ), "MoE only support parallelism on TP or EP dimension."
        
 现在的code看上去，感觉是如果同时设置了TP_size， EP_size,  moe就不工作了。如何能区分启动参数里--tensor-parallel-size  是设给attention的还是MoE的？

File: fastdeploy/model_executor/load_weight_utils.py (hunk 8 above)
Comment (fmiao2372) [reply fastdeploy/model_executor/load_weight_utils.py]: 好的，对V1 loader的支持我们计划后面再做，所以先把该feature相关的code去掉了

File: fastdeploy/engine/args_utils.py (hunk 7 above)
Comment (LeoZhao-BJ) [reply fastdeploy/engine/args_utils.py]: https://github.com/PaddlePaddle/FastDeploy/blame/develop/fastdeploy/model_executor/layers/moe/moe.py#L142

seems latest code already solved our problem.

Comment (fmiao2372) [reply fastdeploy/engine/args_utils.py]: 尝试了这样的配置，目前还有一些问题要解，考虑到还要支持v1 loader，我们先把该feature相关的code去掉，后续改完后会单独再提PR
--tensor-parallel-size = 8
--enable-expert-parallel = true
--data-parallel-size = 1

Comment (yuanlehome) [reply fastdeploy/engine/args_utils.py]: 好的，启动参数配置，可以参考我这个PR描述里的脚本，https://github.com/PaddlePaddle/FastDeploy/pull/4871

**Review by yuanlehome — APPROVED** [review]
approved，正如上文讨论过的那样，我们希望后续可以直接在 V1 loader 下支持新的 feature，再次感谢 intel 的贡献～

**Review by zoooo0820 — COMMENTED** [review]
LGTM

**Review by EmmonsCurse — APPROVED** [review]
